src/lib/y2storage/proposal/space_maker.rb

```diff
@@ -237,57 +269,81 @@ def resize_and_delete(planned_partitions, keep, lvm_helper, disk: nil)
         # restart evaluation
         @distribution = nil
 
-        # maybe it works already...
-        return if success?(planned_partitions)
-
-        # Try various methods to free space. Stop when a valid partition
-        # layout has been found.
-
-        methods_and_args = [
-          # step 1 - resize Windows partitions only if there are no Linux partitions on the disk
-          [:resize_windows!,     [{ force: false }]],
-          # step 2 - # delete Linux partitions
-          [:delete_partitions!,  [:linux, keep]],
-          # step 3 - delete other (non-Windows) partitions
-          [:delete_partitions!,  [:other, keep]],
-          # step 4 - resize Windows partitions
-          #
-          # ** Note **
-          #
-          # There are two steps where we try to resize a Windows partition.
-          # Both should be mutually exclusive. But the 'force' argument
-          # itself doesn't ensure this. However: when we did a resize in
-          # Step 1 we either did a partial resize - in this case we don't
-          # reach this step as it will have freed enough space (else it
-          # would have been a full resize). Or we did a full resize in
-          # Step 1 - then the remaining size in the Windows partition will
-          # be zero and this second resize won't happen.
-          [:resize_windows!,     [{ force: true }]],
-          # step 5 - delete Windows partitions
-          [:delete_partitions!,  [:windows, keep]],
-          # step 6 - if deleting partitions was not enough, maybe there is no
-          # partition table and we have to wipe the disk
-          [:delete_disk_content, [lvm_helper]]
-        ]
-
-        methods_and_args.each_with_index do |method_and_args, idx|
-          method, extra_args = method_and_args
-
-          log.info "Step #{idx + 1} - #{method}"
-          send(method, planned_partitions, disk, *extra_args)
-
-          break if @distribution
+        @disk_name = disk
+        calculate_sorted_candidates(keep)
+
+        until success?(planned_partitions)
+          break unless execute_next_action(planned_partitions, lvm_helper)
         end
 
         raise Error unless @distribution
       end
 
+      # Performs the next action of {#resize_and_delete}
+      #
+      # @return [Boolean] true if some operation was performed. False if nothing
+      #   else could be done to reach the goal.
+      def execute_next_action(planned_partitions, lvm_helper)
+        part = next_partition_to_resize
+        if part
+          target_shrink_size = resizing_size(part, planned_partitions)
+          shrink_partition(part, target_shrink_size)
+        else
+          part_sid = next_partition_to_delete
```

Comment:
I know that `nex_partiton_to_resize` and `next_partition_to_delete` are only internal methods, but for the sake of uniformity I would return the same in both cases, sid or Partition.